### PR TITLE
Create user with short term expiry

### PR DESCRIPTION
**Pull request by negzi**
Register/login user with expiry.

### CODE REVIEW

**Comment by erikjohnston** [comment]
@matrixbot ok to test

(Sorry for matrixbot spam :/)

**Comment by erikjohnston** [comment]
Ah, can you base this PR off of the develop branch please?

File: synapse/rest/client/v1/register.py [hunk 1]
```diff
@@ -355,5 +355,86 @@ def _do_shared_secret(self, request, register_json, session):
...
+    PATTERNS = client_path_patterns("/createUser$", releases=(), include_in_unstable=False)
+
+    def __init__(self, hs):
+        super(CreateUserRestServlet, self).__init__(hs)
+        self.store = hs.get_datastore()
+        self.direct_user_creation_max_duration = hs.config.direct_user_creation_max_duration
+
+    @defer.inlineCallbacks
+    def on_POST(self, request):
+        user_json = _parse_json(request)
```
Comment (Rugvip) [inline synapse/rest/client/v1/register.py]: Use parse_json_object_from_request instead

File: synapse/config/registration.py [hunk 2]
```diff
@@ -53,6 +54,7 @@ def default_config(self, **kwargs):
         # If set, allows registration by anyone who also has the shared
         # secret, even if registration is otherwise disabled.
         registration_shared_secret: "%(registration_shared_secret)s"
+        direct_user_creation_max_duration: 1209600
```
Comment (Rugvip) [inline synapse/config/registration.py]: Put this by itself and with a comment

**Comment by negzi** [comment]
@erikjohnston: Sorry about that I will change the PR to the develop branch ASAP

File: synapse/config/registration.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/config/registration.py]: Might be nice to include units, e.g. `direct_user_creation_max_duration_ms`

File: synapse/handlers/auth.py [hunk 3]
```diff
@@ -438,11 +438,11 @@ def generate_refresh_token(self, user_id):
         ))
         return m.serialize()
 
-    def generate_short_term_login_token(self, user_id):
+    def generate_short_term_login_token(self, user_id, duration_in_sec=(2 * 60 * 1000)):
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: Isn't this milliseconds rather than seconds?

File: synapse/handlers/auth.py [hunk 4]
```diff
@@ -438,11 +438,11 @@ def generate_refresh_token(self, user_id):
...
         macaroon = self._generate_base_macaroon(user_id)
         macaroon.add_first_party_caveat("type = login")
         now = self.hs.get_clock().time_msec()
-        expiry = now + (2 * 60 * 1000)
+        expiry = now + duration_in_sec
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: You're adding seconds and milliseconds

File: synapse/handlers/register.py [hunk 5]
```diff
@@ -361,6 +362,59 @@ def _submit_captcha(self, ip_addr, private_key, challenge, response):
...
+            yield self.check_username(localpart)
+        except SynapseError as e:
+            if e.errcode == Codes.USER_IN_USE:
+                need_register = False
+            else:
+                raise
+
+        user = UserID(localpart, self.hs.hostname)
+        user_id = user.to_string()
+        auth_handler = AuthHandler(self.hs)
```
Comment (erikjohnston) [inline synapse/handlers/register.py]: Use `self.hs.get_handlers().auth_handler`

(Yes, it is a bit sucky that we have to do that, we plan to fix it at some point)

File: synapse/rest/client/v1/register.py [hunk 6]
```diff
@@ -355,5 +355,86 @@ def _do_shared_secret(self, request, register_json, session):
             )
 
 
+class CreateUserRestServlet(ClientV1RestServlet):
+    """Handles user creation via a server-to-server interface
```
Comment (erikjohnston) [inline synapse/rest/client/v1/register.py]: Do you mean the AS interface?

File: synapse/rest/client/v1/register.py [hunk 7]
```diff
@@ -355,5 +355,86 @@ def _do_shared_secret(self, request, register_json, session):
...
+        duration_seconds = 0
+        try:
+            duration_seconds = int(user_json["duration_seconds"])
+        except:
+            raise SynapseError(400, "Failed to parse 'duration_seconds'")
+        if duration_seconds > self.direct_user_creation_max_duration:
+            duration_seconds = self.direct_user_creation_max_duration
+
+        handler = self.handlers.registration_handler
+        (user_id, token) = yield handler.create_user(
```
Comment (erikjohnston) [inline synapse/rest/client/v1/register.py]: You can write `user_id, token = ...`

**Comment by erikjohnston** [comment]
Hi, thanks for the PR! :)

What's the intended semantics for the new `/createUser` endpoint? Is it for AS's to register and relogin users with temporary access tokens? Does the usual `/register` not let you do this (I'm not actually sure how the AS stuff works w.r.t. registration)

**Comment by negzi** [comment]
Hi Erik, you are welcome :)
Actually the intention is to register the user if doesn't exist or login the user and generate a short term token with expiry. 

I will edit your based on your comments also I have made some other fixes based on Patrik's comments.
Is it OK to abandon this PR and make a new one with fixes and publish it on the develop branch?

**Comment by negzi** [comment]
Actually this interface can be used from server to server. While ago Patrik explained it in vector, I will paste it here for more clarification. 

"think an online game where all users already exist in a database, and you want to use synapse for chat
so the game login server just calls the AS to get an access token for a user with a matching id, and you're kinda done
user 'foo' logs into game, login service calls AS and asks for an access token for '@player_foo:mygame.com', and returns that to the user
in that case it doesn't really make sense to have a separate login integration that makes synapse call the game server, it's easier to do it the other way around”

**Comment by erikjohnston** [comment]
> Actually the intention is to register the user if doesn't exist or login the user and generate a short term token with expiry.

Ah, cool, but the API is only hit by application services (since it requires an AS token to be provided)? Is it the intention for the created users to be used by the AS or to be given to actual people?

> I will edit your based on your comments also I have made some other fixes based on Patrik's comments.
> Is it OK to abandon this PR and make a new one with fixes and publish it on the develop branch?
 Hide all checks


Yup, that's fine :)

**Comment by erikjohnston** [comment]
> "think an online game where all users already exist in a database, and you want to use synapse for chat
so the game login server just calls the AS to get an access token for a user with a matching id, and you're kinda done
user 'foo' logs into game, login service calls AS and asks for an access token for '@player_foo:mygame.com', and returns that to the user
in that case it doesn't really make sense to have a separate login integration that makes synapse call the game server, it's easier to do it the other way around”

Aaaah, now I understand.

(When people say server to server I tend to think of "matrix server" to "matrix server" federation traffic)

**Comment by negzi** [comment]
Great!

File: synapse/handlers/register.py (hunk 5 above)
Comment (negzi) [inline synapse/handlers/register.py]: do you mean this way 
auth_handler = AuthHandler(self.hs.get_handlers().auth_handler)

File: synapse/handlers/auth.py (hunk 3 above)
Comment (negzi) [inline synapse/handlers/auth.py]: Oops, my bad :/

File: synapse/handlers/register.py (hunk 5 above)
Comment (erikjohnston) [inline synapse/handlers/register.py]: `self.hs.get_handlers().auth_handler` is a premade AuthHandler :)

So just: `auth_handler = self.hs.get_handlers().auth_handler`

**Pull request closed without merging**